dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preprocess/EventProgramPreProcessor.java

```diff
@@ -66,6 +67,21 @@ public void process( TrackerBundle bundle )
                 ProgramStage programStage = bundle.getPreheat().get( ProgramStage.class, event.getProgramStage() );
                 if ( Objects.nonNull( programStage ) )
                 {
+                    // Program stages should always have a program! Due to how
+                    // metadata import is currently implemented
+                    // it's possible that users run into the edge case that a
+                    // program stage does not have an associated
+                    // program. Tell the user it's an issue with the metadata
+                    // and not the event itself. This should be
+                    // fixed in the metadata import. For more see
+                    // https://jira.dhis2.org/browse/DHIS2-12123
+                    if ( programStage.getProgram() == null )
+                    {
+                        throw new IllegalStateException(
+                            MessageFormat.format(
+                                "Program stage `{0}` has no reference to a program. Check the program stage (DHIS2-12123)",
```

Comment:
For the error message, I'm not sure we should return the jira ticket number; We can rather just keep it plain "Check the program stage configuration" for example, then they in theory know where to look to solve the issue :) Otherwise looks great!